### PR TITLE
3DS and Wii basic platform support

### PR DESCRIPTION
**Pull request by ds-sloth**
I'm putting this code here so we can track it, easily merge main branch changes into it, and have some public review of the included features. Once the Wii code builds on the CI, it should be fully working (but slow to load) even with an unconverted PC asset pack.

Features in this branch:
- [x] 3DS platform support including custom renderer and audio API
- [x] Wii platform support including custom renderer, mixed native and PNG/GIF assets, and full MixerX API
- [x] CLI platform support for extremely fast game logic testing and profiling (can be used in the CI eventually)
- [x] Performance tracking available in the debug info screen
- [x] Simplified player draw code, which adds support for minified spritesheets if graphics are pre-converted

My TODOs before merge:
- [x] Small fix for the player warp draw code with minified spritesheet
- [x] Improve the GIF mask rendering on Wii
- [x] Add SDL2 and MixerX builds to the repo
- [x] Check the GIF image size checker for the CLI target and fix if needed
- [x] Organize the scripts -- place `gfx_convert_wii.py` and `gfx_convert_3ds.py` somewhere other than the repo root
- [x] Cleanup any unnecessary differences between this branch and the main branch, and move the `src/3ds` folder into `core`.
- [x] Do a full review of the code diff (I and Wohlstand should do this, and also any other interested developers) to ensure there are no regressions.
- [x] @Wohlstand, feel free to add any other known issues or TODOs here

Work for me outside the scope of this PR:
- [ ] Before 1.3.7 release, add components with this logic to Moondust Maintainer
- [ ] Merge multires to make the 3DS port playable. Whenever this is merged, I'll update multires with a few hooks needed to make it work well with the Wii and 3DS renderers.
- [ ] Improve spc2it to produce looped 8-channel IT files that do not lose the first ~20ms of sound like they currently do. Also make a converter for any MixerX format to a looped ogg file. These will significantly improve the performance of the game on Old 3DS.

### CODE REVIEW

File: lib/Graphics/graphics_funcs.cpp [hunk 1]
```diff
@@ -201,6 +201,12 @@ void GraphicsHelps::getMaskFromRGBA(FIBITMAP *&image, FIBITMAP *&mask)
                                FreeImage_GetGreenMask(image),
                                FreeImage_GetBlueMask(image));
 
+    if(!mask)
+    {
+        pLogCritical("OOM when extracting mask!");
```
Comment (Wohlstand) [inline lib/Graphics/graphics_funcs.cpp:206]: Please write fully "Out of memory", just make such messages more obvious for newbies who attempt to read the log files.

Comment (ds-sloth) [reply lib/Graphics/graphics_funcs.cpp]: Done, thank you!

File: src/control/input_3ds.cpp [hunk 2]
```diff
@@ -0,0 +1,1162 @@
...
+ KEYID_TOUCH=20,
+ KEYID_CSTICK_RIGHT=24,
+ KEYID_CSTICK_LEFT=25,
+ KEYID_CSTICK_UP=26,
+ KEYID_CSTICK_DOWN=27,
+ KEYID_PAD_RIGHT=28,
+ KEYID_PAD_LEFT=29,
+ KEYID_PAD_UP=30,
+ KEYID_PAD_DOWN=31
+};
```
Comment (Wohlstand) [inline src/control/input_3ds.cpp:68]: Keep indents with 4 spaces :eyes:

File: src/control/input_3ds.cpp [hunk 3]
```diff
@@ -0,0 +1,1162 @@
...
+    }
+    else if(c == ControlsClass::Hotkey)
+    {
+        keys = this->m_hotkeys;
+        keys2 = this->m_hotkeys2;
+        key_max = Hotkeys::n_buttons;
+    }
+    else
+    {
+        // BAD!
```
Comment (Wohlstand) [inline src/control/input_3ds.cpp:485]: Possibly here you could add the log string (use log functions with `D_` prefix if you want to keep these entries printed during debug only, they never appear at release builds)

File: src/control/input_3ds.cpp [hunk 4]
```diff
@@ -0,0 +1,1162 @@
...
+        keys2 = this->m_editor_keys2;
+    }
+    else if(c == ControlsClass::Hotkey)
+    {
+        keys = this->m_hotkeys;
+        keys2 = this->m_hotkeys2;
+    }
+    else
+    {
+        // BAD!
```
Comment (Wohlstand) [inline src/control/input_3ds.cpp:569]: Log? (again, for debug-only use `D_` log calls)

**Review by Wohlstand — CHANGES_REQUESTED** [review]
This is a partial review, I'll give more comments later.

**Comment by Wohlstand** [comment]
Also, I think, I really want to avoid making "MixPlatfor" submodule, and just keeping regular MixerX API for everything: MixerX is already very universal thing, and I don't like the idea of "Universal over universal" :zany_face:

File: src/control/input_3ds.cpp (hunk 3 above)
Comment (ds-sloth) [reply src/control/input_3ds.cpp:485]: I'll do this, thanks!

File: cmake/build_props.cmake [hunk 5]
```diff
@@ -48,6 +48,14 @@ elseif(NOT MSVC)
...
             set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -funwind-tables")
             set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -funwind-tables")
+        elseif(3DS)
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -g -Wl,--gc-sections")
+            set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -g -Wl,--gc-sections")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -g -Wl,--gc-sections")
+        elseif(WII)
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -g -Wl,--gc-sections")
+            set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -g -Wl,--gc-sections")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -g -Wl,--gc-sections")
```
[the remark this answers is not in the document]
Comment (ds-sloth) [reply cmake/build_props.cmake:58]: Did that, thanks!

File: lib/pge_delay.h [hunk 6]
```diff
@@ -4,6 +4,14 @@
...
 #   include <emscripten.h>
 #   define PGE_Delay(x) emscripten_sleep(x)
+#elif defined(__3DS__)
+#   include <3ds.h>
+#   define PGE_Delay(x) svcSleepThread((uint64_t)(x) * 1000000)
+#elif defined(__WII__)
+extern "C" void udelay(unsigned us);
+#   define PGE_Delay(x) udelay((x) * 1000)
+#elif defined(NO_SDL)
+#   define PGE_Delay(x) UNUSED(x)
```
[the remark this answers is not in the document]
Comment (ds-sloth) [reply lib/pge_delay.h:14]: It doesn't seem to exist on a lot of the bare-metal targets that don't support SDL. Maybe better to start with `!defined(NO_SDL)` with the SDL implementation, then look for UNIX and WIN32 fallbacks?

File: CMakeLists.txt [hunk 7]
```diff
@@ -633,7 +736,29 @@ if(VITA)
...
+
+if(NINTENDO_WII)
+    # Link Wii specific libraries to the generic interface.
+    target_link_libraries(A2XT_Int INTERFACE db)
+    target_link_libraries(A2XT_Int INTERFACE wiiuse)
+    target_link_libraries(A2XT_Int INTERFACE fat)
+    target_link_libraries(A2XT_Int INTERFACE asnd)
+    target_link_libraries(A2XT_Int INTERFACE ogc)
+    target_link_libraries(A2XT_Int INTERFACE vorbisidec ogg)
+endif()
```
[the remark this answers is not in the document]
Comment (ds-sloth) [reply CMakeLists.txt:750]: Maybe this is why the in-tree MixerX build wasn't working? I'll try this!

File: lib/DirManager/dirman_posix.cpp [hunk 8]
```diff
@@ -33,14 +33,21 @@ DEALINGS IN THE SOFTWARE.
...
 
 static std::mutex g_dirManMutex;
 
-#define PUT_THREAD_GUARD() \
+#   define PUT_THREAD_GUARD() \
     std::lock_guard<std::mutex> guard(g_dirManMutex);\
     (void)guard;
 
+#else
+#   define PUT_THREAD_GUARD() {}
```
[the remark this answers is not in the document]
Comment (ds-sloth) [reply lib/DirManager/dirman_posix.cpp:48]: Fixed

File: CMakeLists.txt [hunk 9]
```diff
@@ -212,6 +223,23 @@ if(WIN32)
...
+if(NINTENDO_3DS)
+    set(GME_YM2612_EMU "GENS" CACHE STRING "" FORCE)
+    set(BUILD_SHARED_LIBS OFF CACHE BOOL "" FORCE)
+    set(ENABLE_UBSAN OFF CACHE BOOL "" FORCE)
+    set(BUILD_GME_STATIC ON CACHE BOOL "" FORCE)
+    set(GME_ENABLE_ZLIB OFF CACHE BOOL "" FORCE)
+    set(THEXTECH_ENABLE_AUDIO_FX OFF CACHE BOOL "" FORCE)
+    add_subdirectory(3rdparty/AudioCodecs/libgme)
+elseif(NINTENDO_WII)
+    set(BUILD_SHARED_LIBS OFF CACHE BOOL "" FORCE)
```
[the remark this answers is not in the document]
Comment (ds-sloth) [reply CMakeLists.txt:235]: It was, thanks!

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (ds-sloth) [reply CMakeLists.txt:746]: Still doesn't work. I did add these libraries to `library_SDLMixerX.cmake`, though.

File: src/script/luna/lunaimgbox.cpp [hunk 10]
```diff
@@ -18,7 +18,9 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#ifndef PGE_NO_THREADING
 #include <atomic>
+#endif
```
Comment (Wohlstand) [inline src/script/luna/lunaimgbox.cpp:23]: It seems this header is not used here at all, feel free to completely get rid of it.

File: src/graphics/gfx_draw_player.cpp [hunk 11]
```diff
@@ -18,12 +18,121 @@
...
+//! Get left pixel at the player sprite
+int pfrX(const StdPicture& tx, const Player_t& p)
+{
+    // will use internal flags (tx.flags & PLAYER_MODERN and tx.flags & PLAYER_CUSTOM) in future
+    // if tx.flags & PLAYER_CUSTOM, then will use heap-allocated polymorphic CustomData_t* tx.custom_data,
+    // which will include all frame bounding boxes and offsets
+    if(tx.w == 1000)
+        return ((p.Frame * p.Direction + 49) / 10) * 100;
+    else
+    {
```
Comment (Wohlstand) [inline src/graphics/gfx_draw_player.cpp:38]: Really, better to add these flags now, even if they will not be set yet. The w==1000 is a freaking heuristic...

Comment (Wohlstand) [reply src/graphics/gfx_draw_player.cpp:44]: Okay, for such a block I made the "enable experimental features" zone to don't make this work out of our local debugging builds.

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Wohlstand) [reply CMakeLists.txt:808]: Okay, I moved that piece into `library_SDLMixerX.cmake` with a small tweak, and that just works.

File: src/core/base/render_base.cpp [hunk 12]
```diff
@@ -256,7 +255,7 @@ StdPicture AbstractRender_t::LoadPicture(const std::string &path,
     target.frame_h = static_cast<int>(h);
 
     uint8_t *textura = reinterpret_cast<uint8_t *>(FreeImage_GetBits(sourceImage));
-    XRender::loadTexture(target, w, h, textura, pitch);
+    g_render->loadTexture(target, w, h, textura, pitch);
```
Comment (Wohlstand) [inline src/core/base/render_base.cpp:258]: @ds-sloth, why did you replaced `XRender::` with direct `g_render->` calls here?

File: CMakeLists.txt [hunk 13]
```diff
@@ -780,16 +781,6 @@ if(VITA)
         -DVITA -DNO_SCREENSHOT -DTHEXTECH_FIXED_RES -DFIXED_RES -DLOW_MEM -DTHEXTECH_PRELOAD_LEVELS -DNO_INTPTROC)
 endif()
 
-if(NINTENDO_WII)
-    # Link Wii specific libraries to the generic interface.
-    target_link_libraries(A2XT_Int INTERFACE db)
-    target_link_libraries(A2XT_Int INTERFACE wiiuse)
-    target_link_libraries(A2XT_Int INTERFACE fat)
```
Comment (ds-sloth) [inline CMakeLists.txt]: These three are essential to the game overall (even if we did a soundless build) so I'd rather keep them in the main CMake, the others are sound-related and can definitely go to MixerX.

**Review by Wohlstand — APPROVED** [review]
Okay, I think it's fine to merge now. Anything future we can complete later.

**Comment by Wohlstand** [comment]
> Can GameCube controller support be added for the Wii port?

Possibly it should work out of box. Anyway, first I should obtain one to test out some...

**Comment by ds-sloth** [comment]
Doesn't work out-of-the-box because libogc has totally different APIs for GC controllers and Wiimotes. It's possible to implement GC controller support if there is user demand for it, but I'd rather improve the Wii port's stability in other ways first. We should open a separate issue to track this.

**Pull request merged**